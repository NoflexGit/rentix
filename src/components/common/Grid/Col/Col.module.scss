@import './src/assets/styles/variables';

.component {
  padding: 0 12px;
  flex-grow: 1;
  flex-basis: 0;
  max-width: 100%;
}

@for $i from 1 to 12 {
  .width-#{$i} {
    flex: 0 0 calc((100% / 12) * #{$i});
    max-width: calc((100% / 12) * #{$i});
  }
}

@each $breakpoint in $breakpoints {
  $name: nth($breakpoint, 1);
  $size: nth($breakpoint, 2);
  @media screen and (min-width: $size) {
    @for $i from 1 through 12 {
      .width-#{$name}-#{$i} {
        flex: 0 0 calc((100% / 12) * #{$i});
        max-width: calc((100% / 12) * #{$i});
      }
    }
  }
}


