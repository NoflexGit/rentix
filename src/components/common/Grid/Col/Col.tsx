import React, { FC, useMemo } from "react";
import cx from "classnames";
import styles from "./Col.module.scss";

interface IResponsiveBreakpoints extends Record<string, any> {
  xs?: number | string;
  sm?: number | string;
  md?: number | string;
  lg?: number | string;
}

interface IProps {
  children?: React.ReactNode;
  tag?: keyof JSX.IntrinsicElements;
  className?: string;
  width?: IResponsiveBreakpoints | number | string;
  dataTestId?: string;
}

const createClassName = (props?: IResponsiveBreakpoints | string | number) => {
  const classNames: string[] = [];

  if (typeof props === "object") {
    Object.keys(props).forEach((name) => {
      const prop = props[name];

      if (!prop) return;

      classNames.push(styles[`width-${name}-${prop}`]);
    });
  } else {
    classNames.push(styles[`width-${props}`]);
  }

  return classNames;
};

const Col: FC<IProps> = ({
  tag: Component = "div",
  children,
  className,
  width,
  dataTestId,
}) => {
  const widthClasses = useMemo(() => createClassName(width), [width]);
  const classes = cx(styles.component, ...widthClasses, className);

  return (
    <Component className={classes} data-test-id={dataTestId}>
      {children}
    </Component>
  );
};

export default Col;
