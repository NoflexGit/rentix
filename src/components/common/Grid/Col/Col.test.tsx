import React from 'react';
import { render } from '@testing-library/react';

import Col from './Col';

const testId = 'test-id';

describe('Col', () => {
  describe('Snapshots tests', () => {
    it('should match snapshot', () => {
      const { container } = render(<Col>Example</Col>);

      expect(container).toMatchSnapshot();
    });
  });

  describe('Classes tests', () => {
    it('should set `className` class', () => {
      const className = 'test-className';

      const { container } = render(<Col className={className} />);

      expect(container.firstElementChild).toHaveClass(className);
    });

    it('should set width class', () => {
      const { container } = render(<Col width={{ xs: 12, sm: 5, md: 3 }} />);

      expect(container.firstElementChild).toHaveClass('width-xs-12 width-sm-5 width-md-3');
    });
  });

  it('should set `data-test-id` attribute to col', () => {
    const { getByTestId } = render(<Col dataTestId={testId} />);

    expect(getByTestId(testId).tagName).toBe('DIV');
  });

  it('should set a correct tag', () => {
    const { getByTestId } = render(<Col tag="article" dataTestId={testId} />);

    expect(getByTestId(testId).tagName).toBe('ARTICLE');
  });
});
