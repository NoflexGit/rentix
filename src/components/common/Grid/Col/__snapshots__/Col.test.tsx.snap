// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="component width-undefined"
  >
    Example
  </div>
</div>
`;
